diff --git a/plthook_elf.c b/plthook_elf.c
index 47c0116..fc3f3a0 100644
--- a/plthook_elf.c
+++ b/plthook_elf.c
@@ -66,6 +66,11 @@
 #define RTLD_NOLOAD 0
 #endif
 
+#if !defined __UCLIBC__ && !defined __GLIBC__ && !defined __ANDROID__
+// Assume musl libc
+#define __MUSL__
+#endif
+
 #ifndef __GNUC__
 #define __attribute__(arg)
 #endif
@@ -204,7 +209,7 @@ static int check_elf_header(const Elf_Ehdr *ehdr);
 #endif
 static void set_errmsg(const char *fmt, ...) __attribute__((__format__ (__printf__, 1, 2)));
 
-#if defined __ANDROID__ || defined __UCLIBC__
+#if defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__
 struct dl_iterate_data {
     char* addr;
     struct link_map lmap;
@@ -288,7 +293,7 @@ int plthook_open_by_address(plthook_t **plthook_out, void *address)
 {
 #if defined __FreeBSD__
     return PLTHOOK_NOT_IMPLEMENTED;
-#elif defined __ANDROID__ || defined __UCLIBC__
+#elif defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__
     struct dl_iterate_data data = {0,};
     data.addr = address;
     dl_iterate_phdr(dl_iterate_cb, &data);
@@ -312,7 +317,7 @@ int plthook_open_by_address(plthook_t **plthook_out, void *address)
 
 static int plthook_open_executable(plthook_t **plthook_out)
 {
-#if defined __ANDROID__ || defined __UCLIBC__
+#if defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__ || defined __FreeBSD__
     return plthook_open_shared_library(plthook_out, NULL);
 #elif defined __linux__
     return plthook_open_real(plthook_out, _r_debug.r_map);
@@ -340,8 +345,6 @@ static int plthook_open_executable(plthook_t **plthook_out)
         return PLTHOOK_INTERNAL_ERROR;
     }
     return plthook_open_real(plthook_out, r_debug->r_map);
-#elif defined __FreeBSD__
-    return plthook_open_shared_library(plthook_out, NULL);
 #else
     set_errmsg("Opening the main program is not supported on this platform.");
     return PLTHOOK_NOT_IMPLEMENTED;
@@ -351,7 +354,7 @@ static int plthook_open_executable(plthook_t **plthook_out)
 static int plthook_open_shared_library(plthook_t **plthook_out, const char *filename)
 {
     void *hndl = dlopen(filename, RTLD_LAZY | RTLD_NOLOAD);
-#if defined __ANDROID__ || defined __UCLIBC__
+#if defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__
     int rv;
 #else
     struct link_map *lmap = NULL;
@@ -361,7 +364,7 @@ static int plthook_open_shared_library(plthook_t **plthook_out, const char *file
         set_errmsg("dlopen error: %s", dlerror());
         return PLTHOOK_FILE_NOT_FOUND;
     }
-#if defined __ANDROID__ || defined __UCLIBC__
+#if defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__
     rv = plthook_open_by_handle(plthook_out, hndl);
     dlclose(hndl);
     return rv;
@@ -553,7 +556,7 @@ static int plthook_open_real(plthook_t **plthook_out, struct link_map *lmap)
 
 #if defined __linux__
     plthook.plt_addr_base = (char*)lmap->l_addr;
-#if defined __ANDROID__ || defined __UCLIBC__
+#if defined __ANDROID__ || defined __UCLIBC__ || defined __MUSL__
     dyn_addr_base = (const char*)lmap->l_addr;
 #endif
 #elif defined __FreeBSD__ || defined __sun
